src/gt4py/next/iterator/transforms/fuse_as_fieldop.py

```diff
@@ -0,0 +1,195 @@
+    ...     ),
+    ...     im.ref("inp3", field_type),
+    ... )
+    >>> print(nested_as_fieldop)
+    as_fieldop(λ(__arg0, __arg1) → ·__arg0 + ·__arg1, c⟨ IDimₕ: [0, 1) ⟩)(
+      as_fieldop(λ(__arg0, __arg1) → ·__arg0 × ·__arg1, c⟨ IDimₕ: [0, 1) ⟩)(inp1, inp2), inp3
+    )
+    >>> print(
+    ...     FuseAsFieldOp.apply(
+    ...         nested_as_fieldop, offset_provider={}, allow_undeclared_symbols=True
+    ...     )
+    ... )
+    as_fieldop(λ(inp1, inp2, inp3) → ·inp1 × ·inp2 + ·inp3, c⟨ IDimₕ: [0, 1) ⟩)(inp1, inp2, inp3)
+    """  # noqa: RUF002  # ignore ambiguous multiplication character
+
+    uids: eve_utils.UIDGenerator
+
+    @classmethod
+    def apply(
+        cls,
+        node: itir.Program,
+        *,
+        offset_provider,
+        uids: Optional[eve_utils.UIDGenerator] = None,
+        allow_undeclared_symbols=False,
+    ):
+        node = type_inference.infer(
+            node, offset_provider=offset_provider, allow_undeclared_symbols=allow_undeclared_symbols
+        )
+
+        if not uids:
+            uids = eve_utils.UIDGenerator()
+
+        return cls(uids=uids).visit(node)
+
+    def visit_FunCall(self, node: itir.FunCall):
+        node = self.generic_visit(node)
+
+        if cpm.is_call_to(node.fun, "as_fieldop"):
+            node = _canonicalize_as_fieldop(node)
+
+        if cpm.is_call_to(node.fun, "as_fieldop") and isinstance(node.fun.args[0], itir.Lambda):
+            stencil: itir.Lambda = node.fun.args[0]
+            domain = node.fun.args[1] if len(node.fun.args) > 1 else None
+
+            shifts = trace_shifts.trace_stencil(stencil)
+
+            args: list[itir.Expr] = node.args
+
+            new_args: dict[str, itir.Expr] = {}
+            new_stencil_body: itir.Expr = stencil.expr
+
+            for stencil_param, arg, arg_shifts in zip(stencil.params, args, shifts, strict=True):
+                assert isinstance(arg.type, ts.TypeSpec)
+                dtype = type_info.extract_dtype(arg.type)
+                # TODO(tehrengruber): make this configurable
+                should_inline = isinstance(arg, itir.Literal) or (
+                    isinstance(arg, itir.FunCall)
+                    and (cpm.is_call_to(arg.fun, "as_fieldop") or cpm.is_call_to(arg, "if_"))
+                    and (isinstance(dtype, it_ts.ListType) or len(arg_shifts) <= 1)
+                )
+                if should_inline:
+                    if cpm.is_applied_as_fieldop(arg):
+                        pass
+                    elif cpm.is_call_to(arg, "if_"):
+                        type_ = arg.type
+                        arg = im.op_as_fieldop("if_")(*arg.args)
+                        arg.type = type_
+                    elif isinstance(arg, itir.Literal):
+                        arg = im.op_as_fieldop(im.lambda_()(arg))()
+                    else:
+                        raise NotImplementedError()
+
+                    inline_expr, extracted_args = _inline_as_fieldop_arg(arg, self.uids)
```

Comment:
Just a question: since you're passing the only data member of this class as an argument to `_inline_as_fieldop_arg`, wouldn't it be maybe simpler to make it a method?